python/dask_cudf/dask_cudf/expr/_expr.py

```diff
@@ -0,0 +1,205 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+import functools
+
+from dask_expr._cumulative import CumulativeBlockwise, TakeLast
+from dask_expr._groupby import (
+    DecomposableGroupbyAggregation,
+    GroupbyAggregation,
+)
+from dask_expr._reductions import Var
+from dask_expr._shuffle import DiskShuffle
+
+from dask.dataframe.core import _concat
+
+from dask_cudf.groupby import (
+    _finalize_gb_agg,
+    _groupby_partition_agg,
+    _tree_node_agg,
+)
+
+##
+## Custom expression patching
+##
+
+
+class PatchCumulativeBlockwise(CumulativeBlockwise):
+    @property
+    def _args(self) -> list:
+        return self.operands[:1]
+
+    @property
+    def _kwargs(self) -> dict:
+        # Must pass axis and skipna as kwargs in cudf
+        return {"axis": self.axis, "skipna": self.skipna}
+
+
+CumulativeBlockwise._args = PatchCumulativeBlockwise._args
+CumulativeBlockwise._kwargs = PatchCumulativeBlockwise._kwargs
+
+
+def _takelast(a, skipna=True):
+    if not len(a):
+        return a
+    if skipna:
+        a = a.bfill()
+    # Cannot use `squeeze` with cudf
+    return a.tail(n=1).iloc[0]
+
+
+TakeLast.operation = staticmethod(_takelast)
+
+
+_dx_reduction_aggregate = Var.reduction_aggregate
+
+
+def _reduction_aggregate(*args, **kwargs):
+    result = _dx_reduction_aggregate(*args, **kwargs)
+    if result.ndim == 0:
+        return result.tolist()
+    return result
+
+
+Var.reduction_aggregate = staticmethod(_reduction_aggregate)
+
+
+def _shuffle_group(df, col, _filter, p):
+    from dask.dataframe.shuffle import ensure_cleanup_on_exception
+
+    with ensure_cleanup_on_exception(p):
+        g = df.groupby(col)
+        if hasattr(g, "_grouped"):
+            # Avoid `get_group` for cudf data.
+            # See: https://github.com/rapidsai/cudf/issues/14955
+            keys, part_offsets, _, grouped_df = df.groupby(col)._grouped()
+            d = {
+                k: grouped_df.iloc[part_offsets[i] : part_offsets[i + 1]]
+                for i, k in enumerate(keys.to_pandas())
+                if k in _filter
+            }
+        else:
+            d = {i: g.get_group(i) for i in g.groups if i in _filter}
+        p.append(d, fsync=True)
+
+
+DiskShuffle._shuffle_group = staticmethod(_shuffle_group)
+
+
+class CudfGroupbyAggregation(GroupbyAggregation):
+    def _lower(self):
+        return DecomposableCudfGroupbyAggregation(
+            self.frame,
+            self.arg,
+            self.observed,
+            self.dropna,
+            self.split_every,
+            self.split_out,
+            self.sort,
+            self.shuffle_method,
+            self._slice,
+            *self.by,
+        )
+
+
+class DecomposableCudfGroupbyAggregation(DecomposableGroupbyAggregation):
```

Comment:
`DecomposableCudfGroupbyAggregation` is essentially dask-cudf's "optimized" groupby aggregation. Note that this does not seem to be any faster than dask-expr's algorithm any longer (perhaps groupby-apply behavior is much faster in cudf now?). However, we still use a special code path for now to support aggregations like "collect". (tldr: We may be able to remove the custom dask-cudf groupby-aggregation code path soon)